src/services/rate/rate.service.ts

```diff
@@ -1,18 +1,22 @@
-import axios from 'axios';
-import {IRateService, RateData} from "./rate.service.interface";
+import {IRateService} from "./rate.service.interface";
+import {BaseChain} from "./client/chain";
+import privatebankRateService from "./client/privatebank.rate.service";
+import exchangeapiRateService from "./client/exchangeapi.rate.service";
+import bankGovRateService from "./client/bank.gov.rate.service";
 
-class RateService implements IRateService {
-    private readonly apiUrl: string = 'https://api.privatbank.ua/p24api/pubinfo?json&exchange&coursid=5';
-    private readonly currency: string = 'USD';
+export class RateService implements IRateService {
 
     async getExchangeRate(): Promise<number> {
-        const response = await axios.get<RateData[]>(this.apiUrl);
-        const rateData = response.data.find(currency => currency.ccy === this.currency);
-        if (!rateData) {
-            throw new Error('Currency data not found');
-        }
-        return rateData.buy;
+
+        const privatBankChain = new BaseChain(privatebankRateService)
```

Comment:
the main disadvantage of creating chain of responsibility in rate service method is that with every request the program will create new instances of providers so it can affect performance of your app. You can instantiate chain in rate service constructor so it will be created only when you create an instance of rate service 
that is at the start of the application

Reply:
Oh, yeah, it's so useful comment, thanks a lot